dotnet/CoreLib/Pipeline/DistributedPipelineOrchestrator.cs

```diff
@@ -167,28 +184,30 @@ public override async Task RunPipelineAsync(DataPipeline pipeline, CancellationT
 
     private async Task MoveForwardAsync(DataPipeline pipeline, CancellationToken cancellationToken = default)
     {
-        // Note: the pipeline state is persisted in two places:
-        // * source of truth: in the queue (see the message enqueued)
-        // * async copy: in the container together with files - this can be out of sync and is synchronized on dequeue
-
-        if (pipeline.RemainingSteps.Count == 0)
+        if (pipeline.Complete)
         {
             this.Log.LogInformation("Pipeline '{0}' complete", pipeline.DocumentId);
 
-            // Try to save the pipeline status
-            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken, ignoreExceptions: false).ConfigureAwait(false);
+            // Save the pipeline status. If this fails the system should retry the current step.
+            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
         }
         else
         {
             string nextStepName = pipeline.RemainingSteps.First();
             this.Log.LogInformation("Enqueueing pipeline '{0}' step '{1}'", pipeline.DocumentId, nextStepName);
 
+            // Execute as much logic as possible before writing the new pipeline state to disk,
+            // to reduce the chance of the persisted state to be out of sync.
             using IQueue queue = this._queueClientFactory.Build();
             await queue.ConnectToQueueAsync(nextStepName, QueueOptions.PublishOnly, cancellationToken).ConfigureAwait(false);
-            await queue.EnqueueAsync(ToJson(pipeline), cancellationToken).ConfigureAwait(false);
 
-            // Try to save the pipeline status
-            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken, ignoreExceptions: true).ConfigureAwait(false);
+            // Save the pipeline status to disk.
+            // IMPORTANT: If this fails the system will retry the "next" step stored on disk,
+            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
+
+            // Enqueue a pointer to the pipeline (the entire pipeline doc can be too big to fit)
+            // IMPORTANT: If this fails the state on disk will be ahead, and the system will retry the step before.
+            await queue.EnqueueAsync(ToJson(new DataPipelinePointer(pipeline)), cancellationToken).ConfigureAwait(false);
```

Comment:
Why would the system retry the step before? If the previous step is successful, the message should have been removed from its queue at this point. I might be missing something here.

Reply:
When the execution gets here, the system is still processing the message dispatched. The message is still in the queue, locked. At this point the state on disk is saying "step completed" but if this code triggers an exceptions, it bubbles up to the queue dispatcher who doesn't know what went wrong, so it will just unlock the message and let it retry. The message in the queue though is about the step just completed, that's why we say the state on disk is "ahead", and there is a rollback call (line 96).